lxd/util/resources.go

```diff
@@ -0,0 +1,337 @@
+
+			if t != nil {
+				threads[len(threads)-1].name = line
+			}
+		} else if strings.HasPrefix(line, "model name") {
+			i := strings.Index(line, ":")
+			if i < 0 {
+				return nil, err
+			}
+			i++
+
+			line = line[i:]
+			line = strings.TrimSpace(line)
+
+			if t != nil {
+				threads[len(threads)-1].vendor = line
+			}
+		}
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
+	}
+
+	return threads, err
+}
+
+func parseSysDevSystemCpu() ([]thread, error) {
+	ents, err := ioutil.ReadDir("/sys/devices/system/cpu/")
+	if err != nil {
+		return nil, err
+	}
+
+	threads := []thread{}
+	for _, ent := range ents {
+		entName := ent.Name()
+		if !strings.HasPrefix(entName, "cpu") {
+			continue
+		}
+
+		entName = entName[len("cpu"):]
+		idx, err := strconv.Atoi(entName)
+		if err != nil {
+			continue
+		}
+
+		t := thread{}
+		t.ID = uint64(idx)
+		path := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/core_id", t.ID)
+		coreID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.coreID = uint64(coreID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/physical_package_id", t.ID)
+		sockID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.socketID = uint64(sockID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_cur_freq", t.ID)
+		freq, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequency = uint64(freq / 1000)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", t.ID)
+		freq, err = parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequencyTurbo = uint64(freq / 1000)
+
+		threads = append(threads, t)
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
+	}
+
+	return threads, err
+}
+
+func getThreads() ([]thread, error) {
+	threads, err := parseCpuinfo()
+	if err == nil {
+		return threads, nil
+	}
+
+	threads, err = parseSysDevSystemCpu()
+	if err != nil {
+		return nil, err
+	}
+
+	return threads, nil
+}
+
+func CpuResource() (api.ResourcesCPU, error) {
+	c := api.ResourcesCPU{}
+
+	threads, err := getThreads()
+	if err != nil {
+		return c, err
+	}
+
+	var cur *api.ResourcesCPUSocket
+	c.Total = uint64(len(threads))
+	c.Sockets = append(c.Sockets, api.ResourcesCPUSocket{})
+	for _, v := range threads {
+		if uint64(len(c.Sockets)) <= v.socketID {
+			c.Sockets = append(c.Sockets, api.ResourcesCPUSocket{})
+			cur = &c.Sockets[v.socketID]
+		} else {
+			cur = &c.Sockets[v.socketID]
+		}
+
+		if v.coreID+1 > cur.Cores {
+			cur.Cores++
+		}
+
+		cur.Threads++
+		cur.Name = v.name
+		cur.Vendor = v.vendor
+		cur.Frequency = v.frequency
+		cur.FrequencyTurbo = v.frequencyTurbo
+	}
+
+	return c, nil
+}
+
+func MemoryResource() (api.ResourcesMemory, error) {
+	var buffers uint64
+	var cached uint64
+	var free uint64
+	var total uint64
+
+	mem := api.ResourcesMemory{}
+	f, err := os.Open("/proc/meminfo")
```

Comment:
Fine for now, but I suspect we'll eventually want this to look at the current cgroup.

Reply:
Because of use-cases where `LXD` runs nested?

Reply:
Well, nested will get you meminfo through lxcfs. I was thinking of a setup where the user limits the amount of global resources available to LXD itself through cgroups.

Hasn't really come up yet, so definitely fine for now.